### DISCUSSION TITLE
Keeping the "Taskchampion" name?

### DISCUSSION THREAD

**Discussion by djmitche** [GothenburgBitFactory/taskwarrior#2803] [Ideas]
Work is underway to merge what has been called TaskChampion -- specifically, its generic support for task storage and sync -- into Taskwarrior.

While these will be built into the same application, the interface remains important: other applications may also use this "generic support for task storage and sync" to manage tasks, separately from the Taskwarrior CLI.  So, we will have public names for Rust crates, types, functions, etc. that other code (outside of this GitHub org) relies on.

In a strictly practical sense, I'm a little worried that calling all of those things "Taskwarrior" will get really confusing.  We'll have a taskwarrior Task (C++ class) wrapping a taskwarrior Task (Rust type), intermediated by a taskwarrior Task (C / FFI type).

What if we kept the TaskChampion name for the inner bits?  Then this project is a little easier to explain: we're replacing TDB2 with TaskChampion.

Thoughts? Reactions?

**Comment by smemsh** [comment]
Perhaps it wasn't the right choice to merge it into the TaskWarrior repo in the first place.  Already we use submodules for  libshared, why not have it linked in that way? The "generic support for task storage and sync" bits could be made into its own repository (if not already) and pulled in via git submodule, with a TDB3 interface wrapper in the TW C++ (I'd hope it could be a compile-time choice to use TDB2 or TDB3, but that may not be practical given its role with sync).  If it could be made sufficiently generic/abstract, and is usable by multiple other projects, it seems like it's a separate project anyways and should be named whatever it's named, ie TCDB or what have you.

**Reply by djmitche** [reply]
It seems you're echoing the same feelings I have: that the new storage backend is its own thing, usable by other projects, and thus deserves some amount of distinction from the rest of TW.  Maybe a name is one way to represent that.

The repo containing the source is, I think, a separate concern from the naming.  I experimented with a submodule, but found it just introduced too much overhead coordinating the two repos.  I think that would continue to be the case, with PRs spanning the two repos, for example.  In a way, then, putting the Rust code in the TW repo is a kind of "monorepo".  If it works well, we could consider doing the same for libshared (at least the parts that TW uses).

I don't think the choice of DB can be a compile-time thing, unfortunately.  The semantics of TDB2 (multiple files, making changes by diffing tasks, etc.) are a little too deeply embedded to make that practical.

**Reply by smemsh** [reply]
> The repo containing the source is, I think, a separate concern from the naming

Yes, well for sure not calling the generic parts "TaskWarrior Database" when it serves other users than TaskWarrior seems like a good thing, I'm just saying if the project is generic enough, it shouldn't be part of TaskWarrior codebase anyways and thus would naturally need its own generic name along with the separate codebase. TaskWarrior seems more like a downstream application in that case.  Of course we could put all source in the world in one repo.

> I experimented with a submodule, but found it just introduced too much overhead ... [maybe parts or all of libshared should even be moved in]

The parts TimeWarrior uses in common would still need to be separate so the submodule would exist anyways, unless someone wanted to periodically do a subtree split or something and graft the shared parts into TaskWarrior, some do this instead of submodule.  But it does seem that TCDB belongs in a separate repo logically, just my own take on it, usage questions aside.

However in the absence of another downstream project besides TW utilizing TCDB, a separate repo is less compelling and it could always be split off later along with its history as long as it's in a directory tree of its own.

In any case the easiest way that makes sense for the primary developer should trump most other concerns probably

**Comment by djmitche** [comment]
Ah, for some reason I thought libshared was shared between taskwarrior and taskserver, and didn't think of timewarrior.

Anyway, I'll take your use of "TCDB" as indicating that the "TaskChampion" name still has some use, and leave it for the moment.  It'll be no harder to change later than now, anyway.
